### DISCUSSION TITLE
Optimization Roundup  - Discuss <=12GB VRAM Optimization/settings here!

### DISCUSSION THREAD

**Discussion by Centurion-Rome** [d8ahazard/sd_dreambooth_extension#77] [Q&A]
First, big thanks to all contributors! you have made something impressive!

I hope you could optimize it a tiny little more. I Have Super 2080 8GB VRam and almost made it on Win11:

"Tried to allocate 58.00 MiB (GPU 0; 8.00 GiB total capacity;" :(
--xformers
[v] Don't cache latents
[v] se 8bit Adam
Mixed Precision fp16 and "none" tested

Any chance?

**Comment by d8ahazard** [comment]
So, I have a PR here:

https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/4527

Which would let us use "accelerate launch" to run the app, which I think might be the secret sauce to getting it to work on 8GB with a GPU. You also want to disable training the text encoder. I still need to test this on my 8GB GPU and see if it's possible, but if Shivam's repo can be run on 8GB, then this should work too.

Option B is to use "Use CPU", which will work, but very slowly.

**Comment by d8ahazard** [comment]
Related to:

https://github.com/d8ahazard/sd_dreambooth_extension/issues/13

**Comment by Jonseed** [comment]
I have a 3060 12GB card (and only 12GB of system RAM, with 25GB set aside for my paging file), and I was successful in running dreambooth training this morning for the first time. Here were my settings under Advanced:

![Screenshot 2022-11-10 121602](https://user-images.githubusercontent.com/40897189/201186194-d55786e2-b033-43b2-b321-176fb9d88e17.png)

I am running xformers. I was even able to use a classification dataset, and txt files for labels on my dataset. I did _**not**_ save any checkpoints during training, or generate any previews during training (not sure if that would cause OOM, so more testing is needed there to see if I could enable them).

Let me know if you have any questions about my setup, and I'll try to help.

**Reply by mykeehu** [reply]
Here are my settings, everything runs smoothly for me. The other parameters are set to default.
I also have an RTX 3060 12 GB video card, but I have 32 GB RAM and I don't use xformers

![image](https://user-images.githubusercontent.com/6390413/201202503-672d8a68-3d23-4219-a1c6-899cbe5a17b6.png)
![image](https://user-images.githubusercontent.com/6390413/201202574-7ca4cefc-3c70-43e0-80ce-5c836e860bb9.png)

One more thing: above LR 0.0000025 the object will be too concrete and very difficult to shape with the prompts, and the longer you train it, the stronger the model will be. On the other hand, above 0.000001 it generates a lot of garbage and never seems to find what I want. The default of 0.000005 is so strong that I simply could not change anything in the prompt with negative words. Even at 0.000002 and 0.0000025, it became increasingly difficult due to the large number of steps, but it still responded to the other keywords, but kept the parameters of the original images.

**Reply by Jonseed** [reply]
@mykeehu interesting. I wonder if I could gain some speed by not checking "don't cache latents." And I wonder what "pad tokens" does...

**Reply by mykeehu** [reply]
@Jonseed I haven't yet found out whether the _Pad Tokens_ are good or not, the descriptions contradict each other as to whether they should be switched on or off. 
However, using the _horizontal Flip_ I get better quality results for the same number of steps.
The _Don't cache latents_ option, on the other hand, is clearly written in several descriptions not to be turned on, for example here:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/4436

**Reply by Jonseed** [reply]
I tried it unchecking "don't cache latents" and it still worked for me, so that's good. I think it was a bit faster.

**Reply by drnagel** [reply]
Hi, i tried both cfg and couldnt pass when try to train the first step, i have a rtx 3060 but dont know if maybe is related with that im not using studio drivers or isnt related to that.

![imagen](https://user-images.githubusercontent.com/20272314/201363402-aefb7f54-2e18-4604-9e9a-11b27da9bac3.png)
![imagen](https://user-images.githubusercontent.com/20272314/201363464-b1b46feb-1906-470f-aa17-be6f87e2a102.png)

![imagen](https://user-images.githubusercontent.com/20272314/201363014-c428c207-0f62-4637-a68e-bb153624e92d.png)

Any help will be appreaciate, thanks.

**Reply by Jonseed** [reply]
@drnagel make sure you have everything else closed down on your system, background processes, other tabs in your browser, everything. I think it helps to have a fresh boot of the webui server too.

**Reply by drnagel** [reply]
@Jonseed I only have chrome open with sd, nothing else thats weird, maybe something related with windows 11, or drivers, thats why i said about that im using gaming drivers instead of studio drivers. I will try one more time closing almost everything in task manager.
Btw i didnt try xformers because dont know what is that and where can i turn on.

**Reply by Jonseed** [reply]
I would recommend studio drivers. Also try using Edge instead of Chrome, as it uses less system memory (and maybe VRAM?). 

xformers is a speed boost, and might help with VRAM too, I dunno. You can turn it on with the command line argument --xformers when you start the server, in webui-user.bat. Although you might also have to build it for your particular system setup.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Xformers

**Reply by mykeehu** [reply]
As for xformers, it generates slightly different images with the same parmeters, so if you want to train, I don't recommend it. I don't know how much it can interfere with training, but when generating images [you can see the differences between the images.](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/3967)

**Reply by drnagel** [reply]
Tried with Edge and studio drivers but still the dame error, i have 24gb of ram and 24gb of pagefile.
I'm using a classification dataset of 1500 imagen and i put 1500 in 'total number of classification images', but if i set to 0, the training starts. Maybe is somtething relates yo that.

**Reply by Jonseed** [reply]
I haven't tried that large of a class dataset yet, so it could be that...

**Reply by drnagel** [reply]
Finally got it working enabling xformers, without that i couldnt make it work, even using cpu instead of gpu, crashed at half of steps.

Thanks for all the help @Jonseed

**Reply by Jonseed** [reply]
@drnagel glad you got it working! I wasn't sure if xformers would make a difference. I thought such optimizations were unloaded during training, but maybe not.

**Comment by peterhbg** [comment]
I have a 4GB VRAM GPU, and I would like to know if it was possible to train the model on CPU only, because despite checking "use CPU only", when it's time to make the ckpt file I get a CUDA out of memory error. Is this intended or is it an issue? I want to know just so that I don't keep trying if there is no way :)

**Reply by GhostDragon69** [reply]
you don't have enough ram ( i think) how much ram do you have ? ( not gpu)

**Reply by peterhbg** [reply]
15.9GB, but nevermind, I managed to "fix" the problem with the last version, which added a "generate ckpt" button. I train the model with no generating images or checkpoints during training, and at the last step there is the error, but I think the training progress is saved anyways, so I just generate the ckpt manually after training. Now I only need to check if the program is saving training progress despite the error or if only the training steps number is increasing with no training being made.

**Reply by peterhbg** [reply]
Well apparently the function that errors is exactly the function that saves weights, so, no training for me :(

**Comment by LePrau** [comment]
I was able to train on a 10GB 3080

Documentation of the steps and settings I used can be found here: https://github.com/d8ahazard/sd_dreambooth_extension/discussions/84#discussioncomment-4116619

It seems that generating checkpoint every N steps and preview image every N steps does not clean up memory as thorough as the "final" preview image after training ends. So, traing 100 steps, training another 100 steps, and another 100 steps works, but getting a preview every 100 steps does not (OOM crash after the first checkpoint).

**Comment by d8ahazard** [comment]
No, I disabled the bit that unloaded optimizations before training.

**Comment by ghost** [comment]
Something seems off, before the "Generate ckpt" button was added, i managed to run Dreambooth on 3080 10GB on Linux Mint (Nvidia drivers 520.56.06, CUDA version 11.8), with the recommended VRAM optimizations listed above. Training completed with 1000 steps but OOM crashed at that point - and no ckpt file was generated.

After i saw the button added, i can no longer even begin training. Straight out of VRAM. I've tried all tips i've seen here, to no avail.

**edit to add settings:**
- xformers launch arg enabled
- save a checkpoint off
- generate previews off
- don't cache latents checked
- train text encoder unchecked (tried with and without)
- 8bit adam checked
- gradient checkpointing checked
- mixed precision fp16 (tried "no" as well)

**Version info from starting Stable Diffusion:**
Python 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0]
Commit hash: 98947d173e3f1667eba29c904f681047dea9de90
Installing requirements for Web UI
loading Dreambooth reqs from /evo/stable-diffusion/extensions/sd_dreambooth_extension/requirements.txt
Checking Dreambooth requirements.
WARNING: overwriting existing torch/torchvision installation!
Checking torch and torchvision versions
Dreambooth revision is a984e18462fe0b7f6aa10e5816c2ad591c1b784e
Diffusers version is 0.8.0.dev0.
Torch version is 1.12.1+cu116.
Torch vision version is 0.13.1+cu116.

Launching Web UI with arguments: --autolaunch --xformers --ckpt-dir /evo/sd-resources/checkpoints
Preloading Dreambooth!
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
Using VAE found beside selected model
Loading weights [81761151] from /evo/sd-resources/checkpoints/sd-v1-5-pruned-emaonly.ckpt
Global Step: 840000
Loading VAE weights from: /evo/sd-resources/checkpoints/sd-v1-5-pruned-emaonly.vae.pt
Applying xformers cross attention optimization.
Model loaded.
Loaded a total of 0 textual inversion embeddings.
Embeddings: 
Running on local URL

**Reply by mykeehu** [reply]
@jammynads turn off the don't cache latents option and try that. For me, [these settings work](https://github.com/d8ahazard/sd_dreambooth_extension/discussions/77#discussioncomment-4111378) on an RTX 3060 with 12 GB VRAM and turn on under the Settings to move VAE to RAM.

**Reply by ghost** [reply]
Thanks, but unfortunately no progress: still running out of memory. Your card has 2GB more, that's what probably makes the difference in this case.

I removed the .vae file entirely and tried fp16 with "--precision autocast" because someone had issues otherwise - but that launch argument doesn't seem to matter with my setup.

I'm still not sure why it worked before, and not anymore.

**Reply by mykeehu** [reply]
I'm [looking at the code where the button has been added](https://github.com/d8ahazard/sd_dreambooth_extension/commit/f9900ccb3ce7025d1c89334e320a1c3b255bbd07) and an extra module has been imported (traceback), but would it take up that much memory? Because the functions are only called when you click on the button, so in principle it shouldn't affect the training process.

**Reply by ghost** [reply]
I did a complete reinstall of A1111, because it was a pretty long-standing installation and i wanted to be sure i hadn't screwed anything up. No effect.

I do also get "CUDA_SETUP: WARNING! libcudart.so not found in any environmental path", does this have any effect on Dreambooth? It doesn't seem to affect any other function of Stable Diffusion, as far as i can tell everything works like it should. I'm unsure how to fix that, as i have the latest drivers from official sources (Linux Mint driver manager). I recall seeing this error before so it wasn't introduced by the reinstall.

**Reply by ghost** [reply]
Another try, this time there's a different error: **training: name 'str2optimizer8bit_blockwise' is not defined** but still Out of Memory also.

================================================================================
CUDA_SETUP: WARNING! libcudart.so not found in any environmental path. Searching /usr/local/cuda/lib64...
WARNING: No libcudart.so found! Install CUDA or the cudatoolkit package (anaconda)!
CUDA SETUP: Loading binary /evo/stable-diffusion/venv/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cpu.so...
Caching latents: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:02<00:00,  4.69it/s]
 Scheduler Loaded 
 Allocated: 0.0GB 
 Reserved: 0.0GB 

Steps:   0%|                                                                                                                                  | 0/100 [00:00<?, ?it/s] Exception while training: name 'str2optimizer8bit_blockwise' is not defined 
 Allocated: 6.4GB 
 Reserved: 6.8GB 

Traceback (most recent call last):
  File "/evo/stable-diffusion/extensions/sd_dreambooth_extension/dreambooth/train_dreambooth.py", line 941, in main
    optimizer.step()
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/accelerate/optimizer.py", line 134, in step
    self.scaler.step(self.optimizer, closure)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 338, in step
    retval = self._maybe_opt_step(optimizer, optimizer_state, *args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 285, in _maybe_opt_step
    retval = optimizer.step(*args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/optim/lr_scheduler.py", line 65, in wrapper
    return wrapped(*args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/optim/optimizer.py", line 113, in wrapper
    return func(*args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/bitsandbytes/optim/optimizer.py", line 265, in step
    self.update_step(group, p, gindex, pindex)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/bitsandbytes/optim/optimizer.py", line 506, in update_step
    F.optimizer_update_8bit_blockwise(
  File "/evo/stable-diffusion/venv/lib/python3.10/site-packages/bitsandbytes/functional.py", line 858, in optimizer_update_8bit_blockwise
    str2optimizer8bit_blockwise[optimizer_name][0](
NameError: name 'str2optimizer8bit_blockwise' is not defined
 Exception while training: CUDA out of memory. Tried to allocate 30.00 MiB (GPU 0; 9.76 GiB total capacity; 7.44 GiB already allocated; 15.81 MiB free; 7.54 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF 
 Allocated: 6.7GB 
 Reserved: 7.5GB

**Reply by mykeehu** [reply]
[Here is a video with a few tricks](https://www.youtube.com/watch?v=cP44JLtXIeI), did you see this?

**Reply by ghost** [reply]
Yes, i watched the video and tested running Dreambooth once again. I managed to bring my idle VRAM usage down to 77M, which is as low as i can imagine it going, by disabling all desktop effects and lowering resolution from 1440p 165Hz to 720p 60Hz. This saved over 150M, quite a drop. However...

**Still out of memory.** I'm done tearing my hair out over this, i will patiently wait and hope for a technical solution to squeeze just a bit more optimizations out of this. :) Thanks for the suggestions, videos and hard work on the project to everyone involved in it. Much appreciated!

**Reply by jpe90** [reply]
@jammynads In case you're still interested or anyone else finds this, I also encountered this error and solved it on my linux system by adding `LD_LIBRARY_PATH="/opt/cuda/targets/x86_64-linux/lib:$LD_LIBRARY_PATH`. [More context here](https://jeskin.net/blog/dreambooth-low-vram/#name-str2optimizer8bit_blockwise-is-not-defined)

**Comment by d8ahazard** [comment]
By design, the extension should clear all prior VRAM usage before training, and then restore SD back to "normal" when training is complete.

Similarly, someone somewhere was talking about killing their web browser to save VRAM, but I *think* that the VRAM used by the GPU for stuff like browser and desktop windows comes from "shared" VRAM space, while stuff like computations in Torch is used in "reserved' VRAM. 

Or, in a nutshell, unless you're running something like video editing software, CAD stuff, or a 3D video game, I don't *think* running software will have an effect on OOM. 

Of course, I'm also not a hardware engineer, so I could be completely wrong.

**Reply by LePrau** [reply]
> By design, the extension should clear all prior VRAM usage before training, and then restore SD back to "normal" when training is complete.
> 
> Similarly, someone somewhere was talking about killing their web browser to save VRAM, but I _think_ that the VRAM used by the GPU for stuff like browser and desktop windows comes from "shared" VRAM space, while stuff like computations in Torch is used in "reserved' VRAM.
> 
> Or, in a nutshell, unless you're running something like video editing software, CAD stuff, or a 3D video game, I don't _think_ running software will have an effect on OOM.
> 
> Of course, I'm also not a hardware engineer, so I could be completely wrong.

Sadly, the VRAM used by browser or discord or other software seems to be the same "hardware vram" pool as needed for training. I successfully killed training once by simply opening an image on discord ... 

Windows task manager distinguishes between "GPU RAM" and "dedicated GPU RAM" where latter is the number that matters.

**Reply by ghost** [reply]
Similar experience with Textual Inversion/Hypernetwork training; VRAM usage was close to 95% at some point -> opened a news site and began scrolling -> training halted.

**Reply by sgsdxzy** [reply]
If you have an intel cpu, try plugging your monitor to the integrated graphics for the time being.

**Comment by sgsdxzy** [comment]
Anyone has experience with Colossal AI? https://github.com/hpcaitech/ColossalAI/tree/main/examples/images/diffusion
DeepSpeed training is not supported in native Windows yet (even though developed by Microsoft), while Colossal AI uses similar technique to offload VRAM to RAM, claims to train SD on 8GB, and not stated to not support Windows.

**Reply by sgsdxzy** [reply]
However official builds only support torch 1.12+cu113 yet, and 8bit AdamW pins dreambooth to cu116.

**Comment by d8ahazard** [comment]
Related to #124 

On that note, I've added a "wizard" button that attempts to set the "optimal" settings based on the total amount of available VRAM. It's not perfect, but if anybody wants to contribute their "VRAM Total" and settings used to train without OOM, this would help me to help others easily set params to avoid the most common issue with this bit of software. :D

**Comment by mykeehu** [comment]
I'd be interested in the EMA training option if it makes the model better. Is it possible that there will be <= 12 GB VRAM cards? I'm going over the limit with 512 MB, so I'm waiting to see if it will be optimized.

**Reply by mykeehu** [reply]
I tried it this morning, if I use cache for the training and disable preview image generation, EMA training works with 12 GB VRAM.

**Reply by d8ahazard** [reply]
Interestingly, I just read a blog on huggingface last night where they tested various training params, and one of the notes was that EMA doesn't seem to do much.

I'm going to try training a few models today with and without it and see if I can see a difference.

**Reply by mykeehu** [reply]
I turned on xformers, which I don't like to use, but it seems to have helped with the training and I got a good result.

**Comment by d8ahazard** [comment]
Hot off the presses!

https://github.com/d8ahazard/sd_dreambooth_extension/discussions/246

**Comment by mykeehu** [comment]
Something that is not in the hundreds. When I had the simpler version of the training, I could start and continue any number of training sessions, without xformers.
Since I started using the classification images feature, I can only train with xformers, and I can only continue training once, the second time I get an OOM error, so even after resetting, there is still something in memory. 🙄🤔

This is what I saw at the end of the training:

Training complete??
 CLEANUP:
 Allocated: 7.3GB
 Reserved: 10.8GB

 Cleanup Complete.
 Allocated: 7.3GB
 Reserved: 10.8GB

Steps: 100%|███████| 1000/1000 [23:12<00:00,  1.39s/it, loss=0.245, lr=1.22e-7]
 Training completed, reloading SD Model.
 Allocated: 0.0GB
 Reserved: 4.1GB

**Reply by mykeehu** [reply]
@d8ahazard  Now I'm sure: if I generate an image on txt2Img with Restore faces and then try to continue training, I get an OOM error. So the face restorer loads into VRAM, and even if I do a memory optimization, it loads back or gets stuck. Is there any way to get rid of this error before training? Then the next time I generate an image on the txt2Img page, it loads again.

**Comment by Saoqq** [comment]
My 5 cents about how I've managed to run it on 3060ti with 8Gb vram

1) Update windows to 22H2 and update WSL2 to latest kernel (`wsl --update` in windows terminal), otherwise you may receive error with memory pinnig.
    * Python script to check if memory pinning works for you https://github.com/huggingface/diffusers/issues/807#issuecomment-1276584125
     * Also I was not able to run training with `memory pinning = false` on 21H2
 2) After update I was receiving error smth about `cuda illegal memory access`
    * After couple of tries I've discovered that disabling `xformers` and unchecking `8Bit Adam` resolves this issue for me
    * Also reinstalling different versions and building xformers manually didn't help in my case
 3) Even after that I was not able to start training, some isue with `expected <some tensor> received HalfTensor` 
    * Unchecking `Train Text Encoder` helped and eventually I'm able to train with 6s/it
    * 6s/it looks a bit slow but I'm done with all that juggling around WSL and Windows, when have some time will try with native linux
    
To sum up, here is step by step guide that worked in my case:

1) update to 22h2, update WSL
2) install CUDA on WSL (I use latest 11.8)
3) clone A1111 SD
4) install Dreambooth extension via UI
5) stop SD, add envs that mentioned in this repo readme (accelerate, reqs, torch command)
6) install deepspeed
7) run: accelerate config, answer questions
8) now you are ready to go, lauch and start training
    * I was not able to create model from existing .ckpt, OOM happens, so I had to use huggingface instead of local model
    * In my case preview images are not created during training. Failing with following exception
    ```
    Exception with the stupid image again: Input type (torch.cuda.FloatTensor) and weight type (torch.cuda.HalfTensor) should be the same                                                | 0/1 [00:00<?, ?it/s]
    ```
    
`webui-user.sh`
```
export COMMANDLINE_ARGS=""
export TORCH_COMMAND="pip install torch==1.12.1+cu116 torchvision==0.13.1+cu116 --extra-index-url https://download.pytorch.org/whl/cu116"
export REQS_FILE="./extensions/sd_dreambooth_extension/requirements.txt"
export DREAMBOOTH_SKIP_INSTALL=True
export ACCELERATE="True"
```

`~/.cache/huggingface/accelerate/default_config.yaml` after running `accelerate config`
```
command_file: null
commands: null
compute_environment: LOCAL_MACHINE
deepspeed_config:
  gradient_accumulation_steps: 1
  offload_optimizer_device: cpu
  offload_param_device: cpu
  zero3_init_flag: false
  zero_stage: 2
distributed_type: DEEPSPEED
downcast_bf16: 'no'
fsdp_config: {}
gpu_ids: null
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
megatron_lm_config: {}
mixed_precision: fp16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_name: null
tpu_zone: null
use_cpu: false
```
Training config:
![image](https://user-images.githubusercontent.com/52101448/203663278-a4d1520f-76fb-457e-abcd-1b9c992f57ae.png)
